TTHAnalysis/python/tools/nanoAOD/lepScaleFactors.py

```diff
@@ -122,19 +143,56 @@ def analyze(self, event):
+            self.out.fillBranch('leptonSF_2lss%s'%var, leptonSF_2lss)
+            self.out.fillBranch('leptonSF_3l%s'%var  , leptonSF_3l)
+            self.out.fillBranch('leptonSF_4l%s'%var  , leptonSF_4l)
+
+        for var in '_el_up,_el_dn,_mu_up,_mu_dn'.split(','):
+            leptonSF_2lss = 1
+            leptonSF_3l   = 1
+            leptonSF_4l   = 1
+            if len(leps) >= 2:
+                leptonSF_2lss = self.getLooseToTight(leps[0],'',year,2) * self.getLooseToTight(leps[1],'',year,2)
+                leptonSF_2lss = leptonSF_2lss * self.getRecoToLoose(leps[0],var,year) * self.getRecoToLoose(leps[1],var,year)
+            if len(leps) >= 3:
+                leptonSF_3l   = self.getLooseToTight(leps[0],'',year,3) * self.getLooseToTight(leps[1],'',year,3) * self.getLooseToTight(leps[2],'',year,3)
+                leptonSF_3l   = leptonSF_3l *  self.getRecoToLoose(leps[0],var,year) * self.getRecoToLoose(leps[1],var,year) * self.getRecoToLoose(leps[2],var,year)
+            if len(leps) >= 4: 
+                leptonSF_4l   = self.getLooseToTight(leps[0],'',year,3) * self.getLooseToTight(leps[1],'',year,3) * self.getLooseToTight(leps[2],'',year,3) * self.getLooseToTight(leps[3],'',year,3)
+                leptonSF_4l   = leptonSF_4l *  self.getRecoToLoose(leps[0],var,year) * self.getRecoToLoose(leps[1],var,year) * self.getRecoToLoose(leps[2],var,year) * self.getRecoToLoose(leps[3],var,year)
+            self.out.fillBranch('leptonSF_2lss%s'%var, leptonSF_2lss)
+            self.out.fillBranch('leptonSF_3l%s'%var  , leptonSF_3l)
+            self.out.fillBranch('leptonSF_4l%s'%var  , leptonSF_4l)
+        
+
+        # lepton scale factors loose to tight errors
+        for var in '_el_loosetotight_up,_el_loosetotight_dn,_mu_loosetotight_up,_mu_loosetotight_dn'.split(','):
+            leptonSF_2lss = 1
+            leptonSF_3l   = 1
+            leptonSF_4l   = 1
+            if len(leps) >= 2:
+                leptonSF_2lss = self.getLooseToTight(leps[0],var,year,2) * self.getLooseToTight(leps[1],var,year,2)
+                leptonSF_2lss = leptonSF_2lss * self.getRecoToLoose(leps[0],'',year) * self.getRecoToLoose(leps[1],'',year)
+            if len(leps) >= 3:
+                leptonSF_3l   = self.getLooseToTight(leps[0],var,year,3) * self.getLooseToTight(leps[1],var,year,3) * self.getLooseToTight(leps[2],var,year,3)
+                leptonSF_3l   = leptonSF_3l *  self.getRecoToLoose(leps[0],'',year) * self.getRecoToLoose(leps[1],'',year) * self.getRecoToLoose(leps[2],'',year)
+            if len(leps) >= 4: 
+                leptonSF_4l   = self.getLooseToTight(leps[0],var,year,3) * self.getLooseToTight(leps[1],var,year,3) * self.getLooseToTight(leps[2],var,year,3) * self.getLooseToTight(leps[3],var,year,3)
+                leptonSF_4l   = leptonSF_4l *  self.getRecoToLoose(leps[0],'',year) * self.getRecoToLoose(leps[1],'',year) * self.getRecoToLoose(leps[2],'',year) * self.getRecoToLoose(leps[3],'',year)
+
```

Comment:
Wouldn't it be cleaner to just have one loop, and have getLooseToTight and getRecoToLoose be slightlysmarter? they already need to be able to understand if it's electrons or muons and up or down.
e.g. for electrons in getRecoToLoose
`var = +1 if '_el_up' in var_str  else (-1 if '_el_dn' in var_str else 0)`
and in getLooseToTight
`var = +1 if '_el_loosetotight_up' in var_str  else (-1 if '_el_loosetotight_dn' in var_str else 0)`